### DISCUSSION TITLE
issue:  useFormContext + useFieldArray

### DISCUSSION THREAD

**Discussion by pieterjandebruyne** [react-hook-form/react-hook-form#9706] [Q&A]
### Version Number

7.41.3

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-get-started-ts-forked-v5py13?file=/src/ThemeBlock.tsx

### Steps to reproduce

1. Go to https://codesandbox.io/s/react-hook-form-usefieldarray-bug-v5py13?file=/src/ThemeBlock.tsx
2. Click on `DELETE` button

Adding and deleting items in an array using useFormContext + useFieldArray seem to give buggy functionality.

When you delete the items from bottom to start (eg. array of 3: delete 3-2-1) works fine
But when you start deleting in a different order, the values are getting mixed.

As seen in the screencast, on delete sometimes partial objects are maintained (eg. only name and no uuid).
Also values from fields and watch seem to contain different data.. this is a small reproduction of what I have in a project, there I have bigger objects within the array and they are getting mixed quite randomly.

https://user-images.githubusercontent.com/9568274/210339898-15d632b5-3091-4d90-a71a-21b0b543683c.mov

### Expected behaviour

Being able to add / delete items to a fieldarray using useFormContext in nested components.

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Edge

### Relevant log output

```shell
No real errors
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
plz reproduce in this codesandbox: https://codesandbox.io/s/react-hook-form-usefieldarray-ssugn as you have some strange implementation detail, such as cache input name:

```tsx
const TextInput = <T extends FieldValues>({
  name
}: ITextInputProperties<T>): JSX.Element => {
  const { register } = useFormContext<T>(); // retrieve all hook methods
  const inputReference = useRef(name);
  return <input {...register(inputReference.current)} />;
};
```

**Comment by pieterjandebruyne** [comment]
> cache input name

What do you mean with this? does this refer to the `inputReference.current` being passed tot he register function? I thought it was required to use `useRef` if you want fields to be focussed on submit errors.

Not sure what the implementation details are that I should avoid, can try to reproduce within a fork of that sandbox but it still has react 16 though.

**Comment by pieterjandebruyne** [comment]
The issue indeed seems to be connected to the `useRef`. I recreated the codepen with the linked codepen:
first I bumped react to 18 and hook-forms to latest:
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-66dypt?file=/src/index.tsx

Then implemented my logic:
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-9ziy63?file=/src/FormBlock.tsx
In this FormBlock file L 34-38 includes a TextInput (with register(ref)) and TextInput2 (with register(name)).

I though I saw somewhere in the docs we had to use `useRef` to focus inputs with errors on submit but that seems to not be the preferred way?

I also did not use the `Controller` component in my example. Would there be any advantages over using a Controller to achieve what I am doing here (formContext + custom input components where I pass the name to)

Bonus: I normally also use `<TextInput2<T>` for improved typechecking on the formfieldnames etc. But I don't think there is a way to do this for the elements within an array (L40)? it works fine for names without index.

Would love some extra feedback on this but I think the issue can be closed (unless it should be working with useRef)

**Comment by leapful** [comment]
@pieterjandebruyne 

The problem is from using `useRef` to store the field name string which is used to register field element with react-hook-form therefore whenever an item from field array is being removed that causes order shifting, it conflicts with form values order, for example:

```
Input A - ref name "test.[0].firstName" => Removed
Input B - ref name "test.[1].firstName"
Input C - ref name "test.[2].firstName"

becomes

Input B - ref name is still "test.[1].firstName" => It should be "test.[0].firstName"
Input C - ref name is still "test.[2].firstName" => It should be "test[1].firstName"

```

There is no documentation of react-hook-form suggests to use `useRef` hook to store field name and field name is a string which is not related to any DOM reference to focus on error.

Just remove the usage `useRef` and use directly prop `name` will solve the issue.
